src/vcRender.cpp

```diff
@@ -837,67 +914,26 @@ bool vcRender_CreateSelectionBuffer(vcState *pProgramState, vcRenderContext *pRe
-    return result;
-
-  uint8_t depthBytes[4] = {};
-  udUInt2 pickLocation = { (uint32_t)(pRenderContext->currentMouseUV.x * pRenderContext->sceneResolution.x), (uint32_t)(pRenderContext->currentMouseUV.y * pRenderContext->sceneResolution.y) };
-#if GRAPHICS_API_OPENGL
-  pickLocation.y = pRenderContext->sceneResolution.y - pickLocation.y - 1; // upside-down
-#endif
-
-  UD_ERROR_IF(!vcFramebuffer_EndReadPixels(pRenderContext->pFramebuffer, pRenderContext->pDepthTexture, pickLocation.x, pickLocation.y, 1, 1, depthBytes), udR_InternalError); // read previous copy
-  UD_ERROR_IF(!vcFramebuffer_BeginReadPixels(pRenderContext->pFramebuffer, pRenderContext->pDepthTexture, pickLocation.x, pickLocation.y, 1, 1, depthBytes), udR_InternalError); // begin copy for next frame read
-
-  // 24 bit unsigned int -> float
-#if GRAPHICS_API_OPENGL || GRAPHICS_API_METAL
-  pRenderContext->previousFrameDepth = uint32_t((depthBytes[3] << 16) | (depthBytes[2] << 8) | (depthBytes[1] << 0)) / ((1 << 24) - 1.0f);
-  //uint8_t stencil = depthBytes[0];
-#else
-  // TODO (EVC-765): validate this byte order for metal
-  pRenderContext->previousFrameDepth = uint32_t((depthBytes[2] << 16) | (depthBytes[1] << 8) | (depthBytes[0] << 0)) / ((1 << 24) - 1.0f);
-  //uint8_t stencil = depthBytes[3];
-#endif
-
-  // fbo state may not be valid (e.g. first read back will be '0')
-  if (pRenderContext->previousFrameDepth == 0.0f)
-    pRenderContext->previousFrameDepth = 1.0f;
-
-epilogue:
-  return result;
-}
-
 void vcRender_RenderScene(vcState *pProgramState, vcRenderContext *pRenderContext, vcRenderData &renderData, vcFramebuffer *pDefaultFramebuffer)
 {
   udUnused(pDefaultFramebuffer);
 
   float aspect = pRenderContext->sceneResolution.x / (float)pRenderContext->sceneResolution.y;
 
-  vcRender_RenderUD(pProgramState, pRenderContext, renderData);
+  vcRender_RenderUDToTextures(pProgramState, pRenderContext, renderData);
 
   bool selectionBufferActive = vcRender_CreateSelectionBuffer(pProgramState, pRenderContext, renderData);
 
   vcGLState_SetDepthStencilMode(vcGLSDM_LessOrEqual, true);
   vcGLState_SetViewport(0, 0, pRenderContext->sceneResolution.x, pRenderContext->sceneResolution.y);
 
-  vcFramebuffer_Bind(pRenderContext->pFramebuffer);
-  vcFramebuffer_Clear(pRenderContext->pFramebuffer, 0x00FF8080);
+  vcRender_OpaquePass(pProgramState, pRenderContext, renderData); // first pass
+  vcRender_VisualizationPass(pProgramState, pRenderContext); // final pass
 
-  vcRender_PresentUD(pProgramState, pRenderContext);
-
-  vcRenderOpaqueGeometry(pProgramState, pRenderContext, renderData);
-  vcRender_AsyncReadFrameDepth(pRenderContext); // note: one frame behind
-
-
-  // Drawing skybox after opaque geometry saves a bit on fill rate.
-  vcRenderSkybox(pProgramState, pRenderContext);
+  // note: assuming pFramebuffer[1] is still bound at this point
```

Comment:
Why not just rebind?

Reply:
Because vcGL does not yet internally manage redundant bindings, so rebinding might* have some performance implications. As this issue (manage redundant bindings) is logged as an upcoming task, your suggestion was going to be implemented at that time anyway - so I'll just do it now.

* Only did some brief research on the topic and couldn't find anything conclusive. Also did not do any profiling.